theseus/core/objective.py

```diff
@@ -337,7 +337,14 @@ def size_variables(self) -> int:
     def size_aux_vars(self) -> int:
         return len(self.aux_vars)
 
-    def error(self) -> torch.Tensor:
+    def error(self, input_data: Optional[Dict[str, torch.Tensor]]=None, also_update: bool=False) -> torch.Tensor:
```

Comment:
So `input_data` is of the form `Dict[str, torch.Tensor]]` when we have this in `TheseusLayer.forward` but here it might be useful for us to add the option to pass a combination of above and a list of `Variables` for convenience.

For instance, evaluating obstacle cost on a given SE2 robot pose. In the current form, you would need to separate the name and data out of the `Variable` (SE2) object first.

Reply:
@mhmukadam  Is this looks okay ?

`def error(self,input_data: Optional[Dict[str, torch.Tensor]] = None or Optional[List[th.Variable]], also_update: bool = False) > torch.Tensor`

I have a doubt regarding this. If this is the case then we need to check for the type of  `input_data` because we need to copy the older values right ? So how can we copy old values if it's a list ?

Reply:
For this, I think we shouldn't use a list but a dictionary, as in `Optional[Union[Dict[str, torch.Tensor], Dict[str, torch.Variable]]]. Then we do similar logic, except when the type is a variable we use/copy-from `input_data[key].data` instead of `input_data[key]`.

Reply:
Actually, thinking a bit more about this, I don't completely understand the use case you have in mind @mhmukadam. To pass a list of variables, you would already have gone through the effort of creating the necessary variables, giving them the same name as your previous variables, and populating them with the data that you want to evaluate your error on. This seems like more effort than just passing a map of name to tensors?

Reply:
@luisenp Makes more sense. 
> For this, I think we shouldn't use a list but a dictionary, as in `Optional[Union[Dict[str, torch.Tensor], Dict[str, torch.Variable]]]`. Then we do similar logic, except when the type is a variable we use/copy-from `input_data[key].data` instead of `input_data[key]`.

I'll wait for @mhmukadam before i add a commit

Reply:
> Here's an example:
> 
> ```
> p1 = Point2(..., 'p1')
> collision = Collision2D(p1, ...)
> ...
> p2 = ... // some operations that give p2, p2.name could be None
> err = collision.error(p2, also_update=false)
> ```
> 
> Wondering if this is too tricky to implement vs how much we need this? Handling name mismatch would be too complicated but if we restrict to the same name then passing Variable vs torch.Tensor is not that different.

Oh, recall that the error we are computing in this PR is the global objective error. Which means that any input needs to map data to the full set of variables appropriately.

Reply:
That sounds good. We can keep this PR independent of that feature and proceed as is.